### PR TITLE
Nicholascassar/at 194/fix loans ui

### PR DESCRIPTION
**Pull request by nscassar**
Loans functionality implemented

### CODE REVIEW

File: example.env [hunk 1]
```diff
@@ -1,6 +0,0 @@
-# Syncing
```
Comment (mfalthaw) [inline example.env]: Change the name of the file to `.env`

File: src/view/components/AddEntry/index.tsx [hunk 2]
```diff
@@ -1,242 +0,0 @@
-import * as React from 'react';
```
Comment (mfalthaw) [inline src/view/components/AddEntry/index.tsx]: Why did you delete all MilkEntry pages?

File: src/view/components/AddEntry/style.ts [hunk 3]
```diff
@@ -1,24 +0,0 @@
-import { StyleSheet } from 'react-native';
```
Comment (mfalthaw) [inline src/view/components/AddEntry/style.ts]: Same as above.

File: src/view/components/EditEntry/index.tsx [hunk 4]
```diff
@@ -1,200 +0,0 @@
-import * as React from 'react';
```
Comment (mfalthaw) [inline src/view/components/EditEntry/index.tsx]: Same as above.

File: src/view/components/EditEntry/style.ts [hunk 5]
```diff
@@ -1,24 +0,0 @@
-import { StyleSheet } from 'react-native';
```
Comment (mfalthaw) [inline src/view/components/EditEntry/style.ts]: Same as above.

File: src/view/components/EntryDetails/index.tsx [hunk 6]
```diff
@@ -1,165 +0,0 @@
-import * as React from 'react';
```
Comment (mfalthaw) [inline src/view/components/EntryDetails/index.tsx]: Same as above.

File: src/view/components/EntryDetails/style.ts [hunk 7]
```diff
@@ -1,24 +0,0 @@
-import { StyleSheet } from 'react-native';
```
Comment (mfalthaw) [inline src/view/components/EntryDetails/style.ts]: Same as above.

File: src/view/pages/Farmer/index.tsx [hunk 8]
```diff
@@ -9,21 +9,8 @@ const createCollect = () => (
   <Collect />
 );
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: I can't comment on line `5`, but remove `Buy` since we don't have that anymore.

File: src/view/pages/Farmer/tabs/Buy/index.tsx [hunk 9]
```diff
@@ -28,7 +28,7 @@ interface OwnStateType {
 }
```
Comment (mfalthaw) [inline src/view/pages/Farmer/tabs/Buy/index.tsx]: You need to rename the folder to `Loan`.

Comment (mfalthaw) [reply src/view/pages/Farmer/tabs/Buy/index.tsx]: Wait why did you rename some stuff to `loan` but also created another `Loan` folder?

File: src/view/navigation/routes.ts [hunk 10]
```diff
@@ -4,9 +4,12 @@ import FarmerPage from '../pages/Farmer';
 import ExportsPage from '../pages/Exports';
 import LoginPage from '../pages/Login';
 import FarmerSearch from '../pages/FarmerSearch';
-import AddEntryPage from '../components/AddEntry';
-import EditEntryPage from '../components/EditEntry';
-import MilkEntryDetails from '../components/EntryDetails';
+import AddMilkEntryPage from '../components/AddMilkEntry';
+import EditMilkEntryPage from '../components/EditMilkEntry';
+import MilkEntryDetailsPage from '../components/MilkEntryDetails';
```
Comment (mfalthaw) [inline src/view/navigation/routes.ts]: Where are those files you're importing? `AddMilkEntry`, `EditMilkEntry`, ..etc?

File: src/view/navigation/routes.ts [hunk 11]
```diff
@@ -45,15 +48,18 @@ export enum Route {
...
+  MILK_ENTRY_DETAILS = 'Milk Entry Details',
+  EDIT_MILK_ENTRY = 'Edit Milk Entry',
+  ADD_LOAN_ENTRY = 'Add Loan Entry',
+  LOAN_ENTRY_DETAILS = 'Loan Entry Details',
+  EDIT_LOAN_ENTRY = 'Edit Loan Entry',
+  EDIT_FARMER = 'Edit Farmer',
+  ADD_FARMER = 'Add Farmer',
+  ADD_EXPORT_ENTRY = 'Add Delivery Entry',
+  EXPORT_ENTRY_DETAILS = 'Delivery Entry Details',
+  EDIT_EXPORT_ENTRY = 'Edit Delivery Entry',
```
Comment (mfalthaw) [inline src/view/navigation/routes.ts]: I'm pretty sure I edited all of these a while back, did you merger master into your branch?

Comment (mfalthaw) [reply src/view/navigation/routes.ts]: None of this make sense. Idk what the hell is happening.

File: src/view/pages/Farmer/index.tsx [hunk 12]
```diff
@@ -9,21 +9,8 @@ const createCollect = () => (
...
-  <Loan
-    farmerName="Bradley the Farmer"
-    totalWeeklyPaymentBalence="5"
-    totalRemainingBalance="35"
-    loanTransactions={demoLoanTransactions}
-  />
+  <Loan />
 );
 
 const demoPurchaseTransactions = [
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: Get rid of this too.

File: src/view/pages/Farmer/tabs/Loan/index.tsx [hunk 13]
```diff
@@ -1,78 +1,123 @@
...
-    },
+    /** A brief summary at the top of the page */
+    const loanDataSummary = [
+      {
+        label: 'Total Loan Balance',
+        value: this.props.farmerLoanBalance,
+        units: 'UGX',
+      },
+      {
+        label: 'Weekly Balance Including Loans',
```
Comment (mfalthaw) [inline src/view/pages/Farmer/tabs/Loan/index.tsx]: This seems to be too long, lets keep it 2-3 word long.

File: src/view/pages/Farmer/tabs/Loan/index.tsx [hunk 14]
```diff
@@ -1,78 +1,123 @@
...
             <CardSummary
-              data={testData}
+              data={loanDataSummary}
             />
           </Row>
           <Row>
             <DataTable
-              headers={['Date', 'Remaining Balance', 'Weekly Payment']}
-              values={this.props.loanTransactions}
+              headers={['Date', 'Loan Amount']}
```
Comment (mfalthaw) [inline src/view/pages/Farmer/tabs/Loan/index.tsx]: `Amount` should be enough.

File: src/view/navigation/routes.ts (hunk 11 above)
Comment (nscassar) [reply src/view/navigation/routes.ts]: Do you mean back to no spaces between the words?

File: src/view/navigation/routes.ts (hunk 10 above)
Comment (nscassar) [reply src/view/navigation/routes.ts]: Added to most recent push

File: example.env (hunk 1 above)
Comment (nscassar) [reply example.env]: Re-added

File: src/view/components/AddEntry/index.tsx (hunk 2 above)
Comment (nscassar) [reply src/view/components/AddEntry/index.tsx]: Files were untracked. They are added now

File: src/view/components/EditEntry/index.tsx (hunk 4 above)
Comment (nscassar) [reply src/view/components/EditEntry/index.tsx]: Same as above

File: src/view/components/EditEntry/style.ts (hunk 5 above)
Comment (nscassar) [reply src/view/components/EditEntry/style.ts]: Same as above

File: src/view/components/EntryDetails/index.tsx (hunk 6 above)
Comment (nscassar) [reply src/view/components/EntryDetails/index.tsx]: Same as above

File: src/view/pages/Farmer/index.tsx (hunk 8 above)
Comment (nscassar) [reply src/view/pages/Farmer/index.tsx]: Removed, along with test data

File: src/view/pages/Farmer/tabs/Loan/index.tsx (hunk 13 above)
Comment (nscassar) [reply src/view/pages/Farmer/tabs/Loan/index.tsx]: I agree, I just couldn't think of a better way to put it. I'll ask for input at the next meeting

File: src/view/pages/Farmer/tabs/Loan/index.tsx (hunk 14 above)
Comment (nscassar) [reply src/view/pages/Farmer/tabs/Loan/index.tsx]: Updated

File: src/view/pages/Farmer/tabs/Buy/index.tsx (hunk 9 above)
Comment (nscassar) [reply src/view/pages/Farmer/tabs/Buy/index.tsx]: I don't understand what you're asking. I'm assuming this has to do with my previous untracked files. If not, could you please clarify?

File: src/view/components/EditLoanEntry/index.tsx [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+      currency: 'UGX',
+    };
+    this.props.updateLoanEntry(newEntry);
+    this.props.navigate(Route.FARMER);
+  }
+
+  /**
+   * Handle entry changes, update local state
+   */
+  private onAmountChange = (newAmount: string) => this.setState(state => ({ amountOfLoan: parseFloat(newAmount) }));
```
Comment (mfalthaw) [inline src/view/components/EditLoanEntry/index.tsx]: We need to validate input here like we did in `AddMilkEntry`.

File: src/view/components/EditLoanEntry/index.tsx [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+  /**
+   * Returns a button with text, color, and onPress callback specified
+   */
+  private renderButton(text: string, color: ButtonColor, onPress: any) {
+    const isInfo = color === 'INFO';
+    const isPrimary = color === 'PRIMARY';
+
+    return (
+      <Col style={Styles.button}>
+        <Button block info={isInfo} primary={isPrimary} onPress={onPress}>
```
Comment (mfalthaw) [inline src/view/components/EditLoanEntry/index.tsx]: The button shouldn't be press-able unless all inputs are valid. This goes back to input validation.

File: src/view/components/EditLoanEntry/index.tsx [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  private formatEditRow(label: string, value: number | string, onChangeText: any) {
+    return (
+      <Grid>
+        <Row>
+          <Col>
+            <Text>{label}</Text>
+          </Col>
+          <Col>
+          <Item>
+            <Input onChangeText={onChangeText} keyboardType={'numeric'}>
```
Comment (mfalthaw) [inline src/view/components/EditLoanEntry/index.tsx]: Here, copy the `Success` and `Error` functionality from `AddLoanEntry`. We want the user to know if the input is valid or not.

File: src/view/components/EditLoanEntry/index.tsx [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        <Row style={Styles.headerRow}>
+          <Text style={Styles.header}>
+            {moment(this.props.loanEntry.datetime, 'ddd MMM DD Y kk:mm:ss ZZ').local().format('MMMM Do YYYY, h:mm:ss a')}
+          </Text>
+        </Row>
+      </Grid>
+    );
+  }
+
+  private formatEditRow(label: string, value: number | string, onChangeText: any) {
```
Comment (mfalthaw) [inline src/view/components/EditLoanEntry/index.tsx]: Here, the field should be `numeric`, check `AddMilkEntry` and follow the example.

File: src/view/components/EditLoanEntry/index.tsx [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+      <View style={Styles.editView}>
+        {this.formatEditRow('Amount (UGX)', this.props.loanEntry.amount, this.onAmountChange)}
+      </View>
+    );
+  }
+
+  /** Render method for EditLoanEntry */
+  public render() {
+    return(
+      <Content padder style={{ backgroundColor: 'white' }}>
```
Comment (mfalthaw) [inline src/view/components/EditLoanEntry/index.tsx]: The color styling should be coming from the `Styles` file, I think I need ti fix that for my files too.

File: src/view/hoc/TabManager/index.tsx [hunk 20]
```diff
@@ -147,8 +147,8 @@ const mapDispatchToProps: MapDispatchToProps<DispatchPropsType, OwnPropsType> =
         element: () => <Loan />,
       },
       {
-        name: 'Buy',
-        element: () => <Buy />,
+        name: 'Loan',
+        element: () => <Loan />,
```
Comment (mfalthaw) [inline src/view/hoc/TabManager/index.tsx]: How come we have two `Loan` tabs?

File: src/view/navigation/routes.ts [hunk 21]
```diff
@@ -121,19 +127,38 @@ export const routesInfo: RouteInfo[] = [
...
-    component: EditEntryPage,
+    component: EditMilkEntryPage,
     type: 'back',
   },
   {
     route: Route.MILK_ENTRY_DETAILS,
     name: 'Milk Entry Details',
-    component: MilkEntryDetails,
+    component: MilkEntryDetailsPage
+    ,
```
Comment (mfalthaw) [inline src/view/navigation/routes.ts]: Extra line.

File: src/view/pages/Farmer/index.tsx [hunk 22]
```diff
@@ -2,47 +2,34 @@ import * as React from 'react';
 import createTabManager from '../../hoc/TabManager';
 import Collect from './tabs/Collect';
 import Loan from './tabs/Loan';
-import Buy from './tabs/Buy';
+// import Buy from './tabs/Buy';
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: Remove the line since we won't be needing this.

File: src/view/pages/Farmer/index.tsx [hunk 23]
```diff
@@ -2,47 +2,34 @@ import * as React from 'react';
...
-  { date:'Jan 16', product: 'Soap', quantity: '2', price: '6.00' },
-  { date:'Jan 15', product: 'Soap', quantity: '1', price: '3.00' },
-];
+// const demoPurchaseTransactions = [
+//   { date:'Jan 19', product: 'Cow Feed', quantity: '2', price: '6.50' },
+//   { date:'Jan 18', product: 'Soap', quantity: '2', price: '6.00' },
+//   { date:'Jan 17', product: 'Cow Feed', quantity: '2', price: '6.50' },
+//   { date:'Jan 16', product: 'Soap', quantity: '2', price: '6.00' },
+//   { date:'Jan 15', product: 'Soap', quantity: '1', price: '3.00' },
+// ];
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: Remove these.

File: src/view/pages/Farmer/index.tsx [hunk 24]
```diff
@@ -2,47 +2,34 @@ import * as React from 'react';
...
-);
+// const createBuy = () => (
+//   <Buy
+//     farmerName="Bradley the Farmer"
+//     currentDayTotal="1001.01"
+//     currentWeekTotal="50"
+//     currentMonthTotal="405.50"
+//     purchaseTransactions={demoPurchaseTransactions}
+//   />
+// );
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: Remove these.

File: src/view/pages/Farmer/index.tsx [hunk 25]
```diff
@@ -58,10 +45,10 @@ export default createTabManager(
...
       element: createLoan,
     },
-    {
-      name: 'Shop',
-      element: createBuy,
-    },
+    // {
+    //   name: 'Shop',
+    //   element: createBuy,
+    // },
```
Comment (mfalthaw) [inline src/view/pages/Farmer/index.tsx]: Remove these.

File: src/view/pages/Farmer/tabs/Buy/index.tsx (hunk 9 above)
Comment (mfalthaw) [reply src/view/pages/Farmer/tabs/Buy/index.tsx]: I guess what I don't understand here is why you're changing instances of `Buy` to `Loan` inside the `Buy` tab?

File: src/store/modules/loan/selectors.ts [hunk 26]
```diff
@@ -64,11 +66,27 @@ export const getFormattedFarmersTransactions = createSelector(
...
+/**
+ * Selector for the farmer loan account balance
+ * Note: This returns loans only from a week before the current date
+ */
+export const getFarmerLoanBalance = createSelector(
   [getFarmersTransactions],
   (loanEntries: LoanEntry[]) => loanEntries.reduce((sum: number, entry: LoanEntry) =>
-    (inLastWeek(entry.datetime)) ? sum +  entry.amount : sum + 0, 0)).toString();
+    (inLastWeek(entry.datetime)) ? sum +  entry.amount : sum + 0, 0).toString(),
+  );
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/store/modules/loan/selectors.ts]: I'm just a bit confused here. There is an extra indentation? Or you want another one?

File: src/store/modules/loan/selectors.ts [hunk 27]
```diff
@@ -64,11 +66,27 @@ export const getFormattedFarmersTransactions = createSelector(
...
+ * Selector for the farmer total balance for the week
+ * Includes loans and dairy transactions
+ * TODO: Functionality for payments as well
+ */
+export const getFarmerTotalBalance = createSelector(
+  [getFarmerLoanBalance, getFarmerDairyBalance],
+  (loanBalance: string, dairyBalance: string) => 
+  ((parseInt(dairyBalance, radix) - parseInt(loanBalance, radix)).toString()),
+);
+
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/store/modules/loan/selectors.ts]: Removed extra line

File: src/store/modules/milk/selectors.ts [hunk 28]
```diff
@@ -100,6 +100,12 @@ export const getFarmerWeeklyBalance = createSelector(
   (milkEntries: MilkEntry[]) => numberFormatter(milkEntries.reduce((sum: number, entry: MilkEntry) =>
     (inLastWeek(entry.datetime)) ? sum + (entry.costPerUnit * entry.amountOfProduct) : sum + 0, 0)).toString());
 
+/** Selector for the weekly balance without number formatting */
+export const getFarmerWeeklyBalanceNoFormat = createSelector(
+  [getFarmersTransactions],
+  (milkEntries: MilkEntry[]) => milkEntries.reduce((sum: number, entry: MilkEntry) =>
+    (inLastWeek(entry.datetime)) ? sum + (entry.costPerUnit * entry.amountOfProduct) : sum + 0, 0).toString());
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/store/modules/milk/selectors.ts]: I went over the extra selectors after our meeting on Wednesday, and he OK'd them

File: src/view/components/AddLoanEntry/index.tsx [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+  createLoanEntry(newEntry: LoanEntry): Promise<string>;
+}
+
+interface StorePropsType {
+  farmer: Farmer;
+  activeTrader: any;
+  activeFarmer: any;
+}
+
+/** AddEntry PropsType */
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/view/components/AddLoanEntry/index.tsx]: Updated comment

File: src/view/components/AddLoanEntry/index.tsx [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+type PropsType = StorePropsType & DispatchPropsType & OwnPropsType;
+
+interface OwnStateType {
+  amountOfLoan: number;
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/view/components/AddLoanEntry/index.tsx]: Changed to loanAmount

File: src/view/components/AddLoanEntry/index.tsx [hunk 31]
```diff
@@ -0,0 +1,214 @@
...
+    let newEntry: LoanEntry = {
+      type: 'loan',
+      datetime: timeNow,
+      toPersonUuid: this.props.activeTrader,
+      fromPersonUuid: this.props.activeFarmer,
+      amount: this.state.amountOfLoan,
+      currency: 'UGX',
+    };
+    this.props.createLoanEntry(newEntry);
+    this.props.navigate(Route.FARMER);
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/view/components/AddLoanEntry/index.tsx]: I still haven't figured out how to get it to go to the Loans tab after editing, but yes after adding a new loan it goes back to the Loans tab

File: src/view/components/AddLoanEntry/index.tsx [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+        <Item success={this.state.validAmount} error={!this.state.validAmount} floatingLabel>
+          <Label>Amount (UGX)</Label>
+          <Input onChangeText={this.onChangeAmount} keyboardType={'numeric'} />
+        </Item>
+      </Form>
+    );
+  }
+
+  /**
+   * Render method for AddEntry
```
[the remark this answers is not in the document]
Comment (nscassar) [reply src/view/components/AddLoanEntry/index.tsx]: Updated

File: src/view/components/EditLoanEntry/index.tsx (hunk 15 above)
Comment (nscassar) [reply src/view/components/EditLoanEntry/index.tsx]: Input validation added

File: src/view/components/EditLoanEntry/index.tsx (hunk 18 above)
Comment (nscassar) [reply src/view/components/EditLoanEntry/index.tsx]: The example I used was EditMilkEntry, and just checking again I can see that EditMilkEntry does not have input validation as well. I will pull a new issue to fix that later. As for this comment, I'm not sure what you want me to change to 'numeric'

File: src/view/components/EditLoanEntry/index.tsx (hunk 17 above)
Comment (nscassar) [reply src/view/components/EditLoanEntry/index.tsx]: Added

File: src/view/components/EditLoanEntry/index.tsx (hunk 16 above)
Comment (nscassar) [reply src/view/components/EditLoanEntry/index.tsx]: Updated. Button to add is now disabled if input is invalid

File: src/view/components/EditLoanEntry/index.tsx (hunk 19 above)
Comment (nscassar) [reply src/view/components/EditLoanEntry/index.tsx]: Updated

File: src/view/hoc/TabManager/index.tsx (hunk 20 above)
Comment (nscassar) [reply src/view/hoc/TabManager/index.tsx]: We don't. We only have one.

File: src/view/navigation/routes.ts (hunk 21 above)
Comment (nscassar) [reply src/view/navigation/routes.ts]: Fixed

File: src/view/pages/Farmer/index.tsx (hunk 22 above)
Comment (nscassar) [reply src/view/pages/Farmer/index.tsx]: I left it as a comment because I changed it to Payments in a newer branch.

File: src/view/pages/Farmer/index.tsx (hunk 23 above)
Comment (nscassar) [reply src/view/pages/Farmer/index.tsx]: Removed

File: src/view/pages/Farmer/index.tsx (hunk 24 above)
Comment (nscassar) [reply src/view/pages/Farmer/index.tsx]: Removed

File: src/view/pages/Farmer/index.tsx (hunk 25 above)
Comment (nscassar) [reply src/view/pages/Farmer/index.tsx]: Removed

File: src/view/pages/Farmer/tabs/Buy/index.tsx (hunk 9 above)
Comment (nscassar) [reply src/view/pages/Farmer/tabs/Buy/index.tsx]: I think I know what happened. I renamed some stuff when I was working on loan and it changed files globally. I thought I tracked all of them down and fixed it but some must have gone unchanged.

File: src/store/modules/loan/selectors.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (mfalthaw) [reply src/store/modules/loan/selectors.ts]: haha there's an extra indentation.
